compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -2769,6 +2805,146 @@ BType getTypeForUnionTypeMembersAssignableToType(BUnionType unionType, BType tar
         }
     }
 
+    BType getTypeIntersectedWithMapType(BMapType mapType, BType targetType, SymbolEnv env) {
+        if (targetType.tag != TypeTags.MAP && targetType.tag != TypeTags.RECORD) {
+            return symTable.semanticError;
+        }
+        if (targetType.tag == TypeTags.MAP) {
+            BType intersectionConstraintType = getTypeIntersection(mapType.constraint,
+                    ((BMapType) targetType).constraint, env);
+            if (intersectionConstraintType == symTable.semanticError) {
+                return symTable.semanticError;
+            }
+            return new BMapType(TypeTags.MAP, intersectionConstraintType, null);
+        }
+        return getTypeIntersectedWithRecordType((BRecordType) targetType, mapType, env);
+    }
+
+    BType getTypeIntersectedWithRecordType(BRecordType recordType, BType targetType, SymbolEnv env) {
+        if (targetType.tag != TypeTags.MAP && targetType.tag != TypeTags.RECORD) {
+            return symTable.semanticError;
+        }
+        if (targetType.tag == TypeTags.MAP) {
+            BRecordType intersectionRecordType = createAnonymousRecord(env);
+            LinkedHashMap<String, BField> fields = new LinkedHashMap<>();
+            for (String key : recordType.fields.keySet()) {
+                BType intersectionKeyType = getTypeIntersection(recordType.fields.get(key).type,
+                        ((BMapType) targetType).constraint, env);
+                if (intersectionKeyType == symTable.semanticError) {
```

Comment:
Shouldn't we check if the field is optional here? 

For example, `map<json>` and `record {|int i; stream<int> j?;|}` has an intersection, right? 

```
record {|int i;|}
```

Reply:
Fixed